org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/apt/dispatch/BaseProcessingEnvImpl.java

```diff
@@ -123,63 +124,57 @@ public LookupEnvironment getLookupEnvironment() {
+			return SourceVersion.RELEASE_17;
 		}
-		try {
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK1_7) {
-				return SourceVersion.valueOf("RELEASE_7"); //$NON-NLS-1$
+		// From here on we can't use constants that may not be yet defined in
+		// minimal required runtime Java version we have to avoid
+		// errors like java.lang.NoSuchFieldError: RELEASE_20
+		if (sourceLevel > VERSION_FOR_MINIMAL_RUNTIME) {
+			try {
+				return SourceVersion.valueOf("RELEASE_" + CompilerOptions.versionFromJdkLevel(sourceLevel)); //$NON-NLS-1$
+			} catch (IllegalArgumentException e) {
+				// handle call on a minimal release we can run on
+				return MINIMAL_REQUIRED_RUNTIME_VERSION;
 			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK1_8) {
-				return SourceVersion.valueOf("RELEASE_8"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK9) {
-				return SourceVersion.valueOf("RELEASE_9"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK10) {
-				return SourceVersion.valueOf("RELEASE_10"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK11) {
-				return SourceVersion.valueOf("RELEASE_11"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK12) {
-				return SourceVersion.valueOf("RELEASE_12"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK13) {
-				return SourceVersion.valueOf("RELEASE_13"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK14) {
-				return SourceVersion.valueOf("RELEASE_14"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK15) {
-				return SourceVersion.valueOf("RELEASE_15"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK16) {
-				return SourceVersion.valueOf("RELEASE_16"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK17) {
-				return SourceVersion.valueOf("RELEASE_17"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK18) {
-				return SourceVersion.valueOf("RELEASE_18"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK19) {
-				return SourceVersion.valueOf("RELEASE_19"); //$NON-NLS-1$
-			}
-			if (this._compiler.options.sourceLevel == ClassFileConstants.JDK20) {
-				return SourceVersion.valueOf("RELEASE_20"); //$NON-NLS-1$
-			}
-		} catch(IllegalArgumentException e) {
-			// handle call on a JDK 6
-			return SourceVersion.RELEASE_6;
+		} else {
+			// to make compiler happy, should never happen
+			throw new IllegalStateException("Invalid JDK source level: " + sourceLevel); //$NON-NLS-1$
 		}
-		// handle call on a JDK 6 by default
-		return SourceVersion.RELEASE_6;
 	}
 
+	/**
+	 * The minimal required Java runtime version (we need to run compiler on)
+	 */
```

Comment:
I've extra defined them right below the code that uses them, so if anyone later would change any of the code it can be seen. Should I move it up?

Reply:
> I've extra defined them right below the code that uses them, so if anyone later would change any of the code it can be seen. Should I move it up?

That's a fair point. But I think this is something we have been consistently following in JDT. So, it will be better up there.

BTW, good to see that we are cashing in on the move to a higher compliance :)